includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -2911,6 +2912,52 @@ public function cancel_authorization( $order ) {
 		];
 	}
 
+	/**
+	 * Cancels uncaptured authorizations on order cancel.
+	 *
+	 * @param int $order_id - Order ID.
+	 */
+	public function cancel_authorizations_on_order_cancel( $order_id ) {
+		$order = new WC_Order( $order_id );
+		if ( null !== $order ) {
+			$intent_id = $this->order_service->get_intent_id_for_order( $order );
+			if ( null !== $intent_id && '' !== $intent_id ) {
+				try {
+					$request = Get_Intention::create( $intent_id );
+					$request->set_hook_args( $order );
+					$intent = $request->send();
+					$charge = $intent->get_charge();
+
+					/**
+					 * Successful but not captured Charge is an authorization
+					 * that needs to be cancelled.
+					 */
+					if ( null !== $charge
+						&& false === $charge->is_captured()
+						&& Intent_Status::SUCCEEDED === $charge->get_status()
+						&& Intent_Status::REQUIRES_CAPTURE === $intent->get_status()
+					) {
+							$request = Cancel_Intention::create( $intent_id );
+							$request->set_hook_args( $order );
+							$intent = $request->send();
+
+							$this->order_service->update_order_status_from_intent( $order, $intent );
+					}
+
+					$this->order_service->set_intention_status_for_order( $order, $intent->get_status() );
+					$order->save();
+				} catch ( \Exception $e ) {
+					$order->add_order_note(
+						WC_Payments_Utils::esc_interpolated_html(
```

Comment:
Oh, that's nice. Since auto-cancelling is not a standard thing, should we add an order not on success as weel - for observability?

Reply:
That's why I want to use `update_order_status_from_intent` above: it creates a cancellation note. The problem is it is tied with the order status logic and extracting it requires more refactoring than is warranted for this change 😅

Reply:
Okay, test coverage on client is around 62%, which is a bit less than needed for confident refactoring. I could leave without the order note on success.

Marking this conversation branch as resolved.